playback/src/player.rs

```diff
@@ -1789,8 +1942,9 @@ impl PlayerInternal {
 
         let (result_tx, result_rx) = oneshot::channel();
 
-        std::thread::spawn(move || {
-            let data = futures_executor::block_on(loader.load_track(spotify_id, position_ms));
+        let handle = tokio::runtime::Handle::current();
+        thread::spawn(move || {
+            let data = handle.block_on(loader.load_track(spotify_id, position_ms));
```

Comment:
Well like I said, this `librespot` branch is becoming quite the rewrite so I make no promises about backwards compatibility. If anything, I can almost promise that code downstream will need to adapt.

In `librespot-player` I also ran into panics with Tokio and futures. I made various changes to the way threads were spun. Maybe you can take cue from the changes in `player.rs` and see if they work for your own implementation.

Reply:
I am marking this one as "resolved" because this is not an issue with `librespot`, whose player is clearly working.

Reply:
I did some debugging and managed to reproduce the issue with a very simple example.

```rust
use std::time::Duration;

use librespot::core::{
    authentication::Credentials, config::SessionConfig, session::Session, spotify_id::SpotifyId,
};
use librespot::playback::{
    audio_backend::{Sink, SinkResult},
    config::PlayerConfig,
    player::Player,
};
use tokio::time::sleep;

struct MySink;

impl Sink for MySink {
    fn start(&mut self) -> SinkResult<()> {
        Ok(())
    }

    fn write(
        &mut self,
        _packet: &librespot::playback::decoder::AudioPacket,
        _converter: &mut librespot::playback::convert::Converter,
    ) -> librespot::playback::audio_backend::SinkResult<()> {
        Ok(())
    }
}

fn main() {
    let runtime = tokio::runtime::Builder::new_multi_thread()
        .enable_all()
        .worker_threads(1)
        .build()
        .unwrap();

    let _enter = runtime.enter();
    futures::executor::block_on(async move {
        let cred = Credentials::with_password("LOGIN", "PASSWORD");
        let session = Session::connect(SessionConfig::default(), cred, None)
            .await
            .unwrap();
        let track = SpotifyId::from_uri("spotify:track:1p9dxTMtG4gZpLp7yzvT54").unwrap();

        let (mut player, _player_event_channel) =
            Player::new(PlayerConfig::default(), session, None, || {
                Box::new(MySink {})
            });

        player.load(track, true, 0);
    });

    let _enter = runtime.enter();
    futures::executor::block_on(async move {
        sleep(Duration::from_secs(5)).await;
    });
}
```
It works fine using `0.3.1` but crash with your branch:

```
thread '<unnamed>' panicked at 'A Tokio 1.x context was found, but it is being shutdown.', /var/home/cassidy/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/time/driver/entry.rs:547:13
stack backtrace:
   0: rust_begin_unwind
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panicking.rs:100:14
   2: core::panicking::panic_display
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panicking.rs:64:5
   3: tokio::time::driver::entry::TimerEntry::poll_elapsed
             at /var/home/cassidy/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/time/driver/entry.rs:547:13
   4: tokio::time::driver::sleep::Sleep::poll_elapsed
             at /var/home/cassidy/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/time/driver/sleep.rs:357:13
   5: <tokio::time::driver::sleep::Sleep as core::future::future::Future>::poll
             at /var/home/cassidy/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/time/driver/sleep.rs:400:22
   6: librespot_core::spclient::SpClient::request::{{closure}}
             at /var/home/cassidy/dev/rust/librespot/core/src/spclient.rs:184:13
   7: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80:19
   8: librespot_core::spclient::SpClient::get_metadata::{{closure}}
             at /var/home/cassidy/dev/rust/librespot/core/src/spclient.rs:206:9
   9: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80:19
  10: librespot_core::spclient::SpClient::get_track_metadata::{{closure}}
             at /var/home/cassidy/dev/rust/librespot/core/src/spclient.rs:210:9
  11: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80:19
  12: <librespot_metadata::track::Track as librespot_metadata::Metadata>::request::{{closure}}
             at /var/home/cassidy/dev/rust/librespot/metadata/src/track.rs:107:9
  13: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80:19
  14: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/future.rs:119:9
  15: librespot_metadata::Metadata::get::{{closure}}
             at /var/home/cassidy/dev/rust/librespot/metadata/src/lib.rs:49:24
  16: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80:19
  17: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/future.rs:119:9
  18: <librespot_metadata::track::Track as librespot_metadata::audio::item::InnerAudioItem>::get_audio_item::{{closure}}
             at /var/home/cassidy/dev/rust/librespot/metadata/src/track.rs:70:21
  19: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80:19
  20: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/future.rs:119:9
  21: librespot_metadata::audio::item::AudioItem::get_file::{{closure}}
             at /var/home/cassidy/dev/rust/librespot/metadata/src/audio/item.rs:34:39
  22: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80:19
  23: librespot_playback::player::PlayerTrackLoader::load_track::{{closure}}
             at /var/home/cassidy/dev/rust/librespot/playback/src/player.rs:741:27
  24: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80:19
  25: tokio::park::thread::CachedParkThread::block_on::{{closure}}
             at /var/home/cassidy/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/park/thread.rs:263:54
  26: tokio::coop::with_budget::{{closure}}
             at /var/home/cassidy/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/coop.rs:106:9
  27: std::thread::local::LocalKey<T>::try_with
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/thread/local.rs:399:16
  28: std::thread::local::LocalKey<T>::with
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/thread/local.rs:375:9
  29: tokio::coop::with_budget
             at /var/home/cassidy/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/coop.rs:99:5
  30: tokio::coop::budget
             at /var/home/cassidy/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/coop.rs:76:5
  31: tokio::park::thread::CachedParkThread::block_on
             at /var/home/cassidy/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/park/thread.rs:263:31
  32: tokio::runtime::enter::Enter::block_on
             at /var/home/cassidy/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/enter.rs:151:13
  33: tokio::runtime::handle::Handle::block_on
             at /var/home/cassidy/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/handle.rs:326:9
  34: librespot_playback::player::PlayerInternal::load_track::{{closure}}
             at /var/home/cassidy/dev/rust/librespot/playback/src/player.rs:1944:24
```

This code is totally safe and so shouldn't panic. It seems that there is some pending async operation that is not properly cleaned up when the player is dropped.

Reply:
Did you try moving away from `futures::executor::block_on` like I did in `player.rs`, have you looked at the diff?

Not at all to be unfriendly, but I don't find it concerning that code working with 0.3.1 isn't working anymore now. So I am curious if my solution for `player.rs` will also benefit you. If not I welcome an extra pair of eyes to see where I am missing a `Drop`.

Reply:
Again the native `librespot` player (binary) is proof that things can run just fine.

Reply:
Using `runtime.block_on()` instead of `futures::executor::block_on()`, as you did for the player, does not help, same crash.

> Again the native librespot player (binary) is proof that things can run just fine.

Sure, but the player is not hitting this kind of race either.  My point is that safe use of the API can result in a crash which is very likely because of a bug in the new async code.

Reply:
See the logs, the player keeps loading the track after `PlayerInternal` has been dropped and its thread terminated:

```
[2021-12-27T22:44:34Z DEBUG librespot_playback::player] Shutting down player thread ...
[2021-12-27T22:44:34Z DEBUG librespot_playback::player] new Player[0]
[2021-12-27T22:44:34Z INFO  librespot_playback::convert] Converting with ditherer: tpdf
[2021-12-27T22:44:34Z INFO  librespot_core::session] Country: "BE"
[/var/home/cassidy/dev/rust/librespot/playback/src/player.rs:938] self.commands.poll_recv(cx) = Ready(
    Some(
        Load(
            SpotifyId(
                "spotify:track:1p9dxTMtG4gZpLp7yzvT54",
            ),
            true,
            0,
        ),
    ),
)
[2021-12-27T22:44:34Z DEBUG librespot_playback::player] command=Load(SpotifyId("spotify:track:1p9dxTMtG4gZpLp7yzvT54"), true, 0)
[/var/home/cassidy/dev/rust/librespot/playback/src/player.rs:938] self.commands.poll_recv(cx) = Ready(
    None,
)
[/var/home/cassidy/dev/rust/librespot/playback/src/player.rs:1944] "load track" = "load track"
[2021-12-27T22:44:34Z DEBUG librespot_playback::player] drop PlayerInternal[0]
[/var/home/cassidy/dev/rust/librespot/metadata/src/track.rs:70] "get_audio_item" = "get_audio_item"
[2021-12-27T22:44:34Z DEBUG librespot_playback::player] PlayerInternal thread finished.
[2021-12-27T22:44:34Z DEBUG librespot::component] new SpClient
[2021-12-27T22:44:34Z INFO  librespot_core::spclient] Resolved "gew1-spclient.spotify.com:443" as spclient access point
[2021-12-27T22:44:34Z DEBUG librespot::component] new TokenProvider
[2021-12-27T22:44:34Z TRACE librespot_core::token] Requested token in scopes "playlist-read" unavailable or expired, requesting new token.
[2021-12-27T22:44:34Z DEBUG librespot::component] new MercuryManager
[2021-12-27T22:44:34Z DEBUG librespot_playback::player] player thread shutted down
[2021-12-27T22:44:34Z TRACE librespot_core::session] Received product info: {

(...)
```

Reply:
So, the root cause of the problem is that the thread created by `PlayerInternal::load_track()` is not joined when `PlayerInternal` is dropped; see [this WIP patch fixing the panic](https://github.com/gdesmott/librespot/commit/9bc49b1088b32dcf092d9ee6e12866df11d2a132).

The patch needs some more work as the handles keep piling up in the vector until we drop but that should be easy enough to fix. I'm happy to finish it if you agree with the fix.

Reply:
Thanks, that's good info - I'm marking this as unresolved. Somehow the other futures called by `PlayerInternal` aren't dropped when `PlayerInternal` goes out of scope.

Reply:
> So, the root cause of the problem is that the thread created by `PlayerInternal::load_track()` is not joined when `PlayerInternal` is dropped; see [this WIP patch fixing the panic](https://github.com/gdesmott/librespot/commit/9bc49b1088b32dcf092d9ee6e12866df11d2a132).
> 
> The patch needs some more work as the handles keep piling up in the vector until we drop but that should be easy enough to fix. I'm happy to finish it if you agree with the fix.

Great stuff! Our replies crossed in time. For sure, please go ahead.

Before you posted your fix I was thinking if we shouldn't be using `select!`. If I'm right I think it handles abort logic a little better, but I'm no expert on that yet.

Reply:
> Before you posted your fix I was thinking if we shouldn't be using `select!`. If I'm right I think it handles abort logic a little better, but I'm no expert on that yet.

I don't think that will do as we need to wait on threads, not futures.

[Here is what I came up with](https://github.com/gdesmott/librespot/commits/race-fix). I would have opened a PR but for some reason Github does not list your fork in the base dropdown menu.

Reply:
> I would have opened a PR but for some reason Github does not list your fork in the base dropdown menu.

@roderickvd cough, cough merge what you have so other can work on it easily also...

Reply:
> [Here is what I came up with](https://github.com/gdesmott/librespot/commits/race-fix). I would have opened a PR but for some reason Github does not list your fork in the base dropdown menu.

Way to go @gdesmott. If we can all agree on the branching & committing workflow I just proposed in the comments, I can merge this, and you can send in a PR and get the proper attribution too.

> > Before you posted your fix I was thinking if we shouldn't be using `select!`. If I'm right I think it handles abort logic a little better, but I'm no expert on that yet.
> 
> I don't think that will do as we need to wait on threads, not futures.

That's true. Somewhere it doesn't sit right with me that we are spawning with `std::thread::spawn` instead of `tokio::spawn` (or `session.spawn` which does the same thing, mostly) which might be doing a little better bookkeeping. But I didn't get it to work fast enough so I went on to other things first. If anyone has any ideas, do chime in.

Reply:
> Way to go @gdesmott. If we can all agree on the branching & committing workflow I just proposed in the comments, I can merge this, and you can send in a PR and get the proper attribution too.

I agree with your proposal. The current mega PR isn't convenient for neither testing or contributing.